File: LLD/Design-Patterns/Mediator/code/Mediator.java
// Mediator Pattern — a ChatRoom routes messages between Users instead of Users talking directly.
// Compile: javac Mediator.java
// Run:     java Mediator

import java.util.ArrayList;
import java.util.List;

interface ChatMediator {
    void sendMessage(String message, User sender);
    void addUser(User user);
}

class ChatRoom implements ChatMediator {
    private List<User> users = new ArrayList<>();

    public void addUser(User user) {
        users.add(user);
    }

    public void sendMessage(String message, User sender) {
        for (User user : users) {
            if (user != sender) {
                user.receive(message, sender.getName());
            }
        }
    }
}

class User {
    private String name;
    private ChatMediator mediator;

    User(String name, ChatMediator mediator) {
        this.name = name;
        this.mediator = mediator;
    }

    String getName() {
        return name;
    }

    void send(String message) {
        System.out.println(name + " sends: " + message);
        mediator.sendMessage(message, this);
    }

    void receive(String message, String fromName) {
        System.out.println(name + " received from " + fromName + ": " + message);
    }
}

public class Mediator {
    public static void main(String[] args) {
        ChatRoom chatRoom = new ChatRoom();

        User alice = new User("Alice", chatRoom);
        User bob = new User("Bob", chatRoom);
        User carol = new User("Carol", chatRoom);

        chatRoom.addUser(alice);
        chatRoom.addUser(bob);
        chatRoom.addUser(carol);

        alice.send("Hey everyone!");
        bob.send("Hi Alice!");
    }
}

File: LLD/Design-Patterns/Mediator/code/Mediator.cpp
// Mediator Pattern — a ChatRoom routes messages between Users instead of Users talking directly.
// Compile: g++ -std=c++14 Mediator.cpp -o mediator
// Run:     ./mediator

#include <iostream>
#include <string>
#include <vector>

class User;

class ChatMediator {
public:
    virtual void sendMessage(const std::string& message, User* sender) = 0;
    virtual void addUser(User* user) = 0;
    virtual ~ChatMediator() = default;
};

class User {
public:
    User(std::string name, ChatMediator* mediator) : name_(std::move(name)), mediator_(mediator) {}

    const std::string& getName() const { return name_; }

    void send(const std::string& message) {
        std::cout << name_ << " sends: " << message << std::endl;
        mediator_->sendMessage(message, this);
    }

    void receive(const std::string& message, const std::string& fromName) {
        std::cout << name_ << " received from " << fromName << ": " << message << std::endl;
    }

private:
    std::string name_;
    ChatMediator* mediator_;
};

class ChatRoom : public ChatMediator {
public:
    void addUser(User* user) override {
        users_.push_back(user);
    }

    void sendMessage(const std::string& message, User* sender) override {
        for (auto* user : users_) {
            if (user != sender) {
                user->receive(message, sender->getName());
            }
        }
    }

private:
    std::vector<User*> users_;
};

int main() {
    ChatRoom chatRoom;

    User alice("Alice", &chatRoom);
    User bob("Bob", &chatRoom);
    User carol("Carol", &chatRoom);

    chatRoom.addUser(&alice);
    chatRoom.addUser(&bob);
    chatRoom.addUser(&carol);

    alice.send("Hey everyone!");
    bob.send("Hi Alice!");

    return 0;
}
